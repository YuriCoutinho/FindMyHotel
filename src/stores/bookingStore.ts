import { defineStore } from 'pinia'

export const useBookingStore = defineStore('bookingStore', {
  state: () => ({
    selectedHotel: null as any | null,
  }),
  getters: {
    getSelectedHotel: (state) => state.selectedHotel,
  },
  actions: {
    setSelectedHotel(hotel: any) {
      this.selectedHotel = hotel
    }
  }
})
