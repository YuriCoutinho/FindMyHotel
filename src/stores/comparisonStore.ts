import { defineStore } from 'pinia'

export const useComparisonStore = defineStore('comparisonStore', {
  state: () => ({
    hotelsToCompare: [] as any[]
  }),
  actions: {
    addHotel(hotel: any) {
      if (!this.hotelsToCompare.find((h) => h.id === hotel.id)) {
        this.hotelsToCompare.push(hotel)
      }
    },
    removeHotel(hotelId: string) {
      this.hotelsToCompare = this.hotelsToCompare.filter((hotel) => hotel.id !== hotelId)
    }
  }
})
