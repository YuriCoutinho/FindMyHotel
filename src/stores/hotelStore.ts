import { defineStore } from 'pinia'
import axios from 'axios'
import { hotels as initialHotels } from '@/utils/constants'

export const useHotelStore = defineStore('hotelStore', {
  state: () => ({
    selectedState: null as { name: string; abbreviation: string } | null,
    selectedCity: '' as string | null,
    cities: [] as string[],
    checkInDate: '' as string | null,
    checkOutDate: '' as string | null,
    guests: 1,
    rooms: 1,
    searchIsLoading: false as boolean,
    hotels: [] as any[],
    originalHotels: [] as any[]
  }),
  getters: {
    getSelectedState: (state) => state.selectedState,
    getSelectedCity: (state) => state.selectedCity,
    getCheckInDate: (state) => state.checkInDate,
    getCheckOutDate: (state) => state.checkOutDate,
    getGuests: (state) => state.guests,
    getRooms: (state) => state.rooms,
    getSearchIsLoading: (state) => state.searchIsLoading,
    getHotels: (state) => state.hotels,
    getCities: (state) => state.cities,
    isCitySelectionEnabled: (state) => state.cities.length > 0,
    isFormValid: (state) => {
      return (
        state.selectedState !== null &&
        state.selectedCity !== '' &&
        state.checkInDate !== '' &&
        state.checkOutDate !== '' &&
        state.guests > 0 &&
        state.rooms > 0
      )
    }
  },
  actions: {
    setState(state: { name: string; abbreviation: string } | null) {
      this.selectedState = state
      if (state) {
        this.searchCities(state.abbreviation)
      }
    },
    setCity(city: string | null) {
      this.selectedCity = city
    },
    setCheckInDate(checkInDate: string | null) {
      this.checkInDate = checkInDate
    },
    setCheckOutDate(checkOutDate: string | null) {
      this.checkOutDate = checkOutDate
    },
    setGuests(selectedGuests: number) {
      this.guests = selectedGuests
    },
    setRooms(selectedRooms: number) {
      this.rooms = selectedRooms
    },
    setSearchIsLoading(searchIsLoading: boolean) {
      this.searchIsLoading = searchIsLoading
    },
    setHotels(hotels: any[]) {
      this.hotels = hotels
    },
    sortHotelsByPriceAsc() {
      const sortedHotels = [...this.hotels].sort((a, b) => a.price - b.price)
      this.setHotels(sortedHotels)
    },
    sortHotelsByPriceDesc() {
      const sortedHotels = [...this.hotels].sort((a, b) => b.price - a.price)
      this.setHotels(sortedHotels)
    },
    sortHotelsByStarsAsc() {
      const sortedHotels = [...this.hotels].sort((a, b) => a.stars - b.stars)
      this.setHotels(sortedHotels)
    },
    sortHotelsByStarsDesc() {
      const sortedHotels = [...this.hotels].sort((a, b) => b.stars - a.stars)
      this.setHotels(sortedHotels)
    },
    resetHotels() {
      this.setHotels([...this.originalHotels])
    },
    async searchCities(stateAbbreviation: string) {
      this.selectedCity = null
      this.cities = []
      if (stateAbbreviation) {
        try {
          const response = await axios.get(
            `https://servicodados.ibge.gov.br/api/v1/localidades/estados/${stateAbbreviation}/municipios`
          )
          this.cities = response.data.map((municipality: any) => municipality.nome)
        } catch (error) {
          console.error('Error fetching municipalities:', error)
          this.cities = []
        }
      }
    },
    async searchHotels() {
      this.setSearchIsLoading(true)
      try {
        setTimeout(() => {
          this.hotels = [...initialHotels]
          this.originalHotels = [...initialHotels]
          this.setSearchIsLoading(false)
        }, 3000)
      } catch (error) {
        console.error('Error searching hotels:', error)
        this.setSearchIsLoading(false)
      }
    }
  }
})
