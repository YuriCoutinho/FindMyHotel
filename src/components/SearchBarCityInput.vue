<template>
  <AutoComplete
    v-model="selectedCity"
    :suggestions="filteredCities"
    :disabled="!enabled"
    inputClass="w-full md:w-40 text-gray-700 font-medium border rounded-lg px-3 py-2 focus:outline-none focus:ring-1 focus:ring-emerald-500"
    panelClass="bg-white shadow-2xl rounded-lg mt-1 p-3 overflow-y-auto cursor-pointer"
    placeholder="Cidade de destino"
    forceSelection
    @complete="search"
    @blur="onBlur"
  />
</template>

<script setup lang="ts">
import { ref, computed } from 'vue'
import AutoComplete from 'primevue/autocomplete'

import { useHotelStore } from '@/stores/hotelStore'

const hotelStore = useHotelStore()

const query = ref<string>('')

const enabled = computed(() => hotelStore.isCitySelectionEnabled)

const filteredCities = computed(() => {
  if (hotelStore.getCities.length === 0) return []
  return hotelStore.getCities.filter((city) =>
    city.toLowerCase().startsWith(query.value.toLowerCase())
  )
})

const selectedCity = computed({
  get: () => hotelStore.getSelectedCity,
  set: (value) => hotelStore.setCity(value)
})

const search = (event: { query: string }) => {
  query.value = event.query
}

const onBlur = () => {
  if (selectedCity.value) {
    hotelStore.setCity(selectedCity.value)
  }
}
</script>
