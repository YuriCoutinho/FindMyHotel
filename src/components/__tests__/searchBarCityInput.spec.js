import { mount } from '@vue/test-utils'
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { createPinia, setActivePinia } from 'pinia'
import PrimeVue from 'primevue/config'
import AutoComplete from 'primevue/autocomplete'
import SearchBarCityInput from '@/components/SearchBarCityInput.vue'
import { useHotelStore } from '@/stores/hotelStore'

describe('SearchBarCityInput.vue', () => {
  let hotelStore

  beforeEach(() => {
    const pinia = createPinia()
    setActivePinia(pinia)
    hotelStore = useHotelStore()

    // Mocking hotelStore data and methods using $patch
    hotelStore.$patch({
      cities: ['New York', 'Los Angeles', 'Chicago'],
      selectedCity: ''
    })

    vi.spyOn(hotelStore, 'setCity')
  })

  const createWrapper = () => {
    return mount(SearchBarCityInput, {
      global: {
        plugins: [PrimeVue],
        components: { AutoComplete }
      }
    })
  }

  it('Should render the component', () => {
    const wrapper = createWrapper()
    expect(wrapper.exists()).toBe(true)
  })

  it('Should initialize with the correct value', () => {
    const wrapper = createWrapper()
    expect(wrapper.findComponent(AutoComplete).props('modelValue')).toBe(hotelStore.selectedCity)
  })

  it('Should match the snapshot', () => {
    const wrapper = createWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Should filter cities based on search query', async () => {
    const wrapper = createWrapper()
    const autoCompleteComponent = wrapper.findComponent(AutoComplete)

    autoCompleteComponent.vm.$emit('complete', { query: 'New' })
    await wrapper.vm.$nextTick()

    // Accessing the filtered cities through the store
    const filteredCities = hotelStore.cities.filter((city) => city.toLowerCase().startsWith('new'))
    expect(filteredCities.length).toBe(1)
    expect(filteredCities).toContain('New York')
  })

  it('Should update the store when a city is selected', async () => {
    const wrapper = createWrapper()
    const autoCompleteComponent = wrapper.findComponent(AutoComplete)

    await autoCompleteComponent.vm.$emit('update:modelValue', 'Chicago')
    await wrapper.vm.$nextTick()

    expect(hotelStore.setCity).toHaveBeenCalledWith('Chicago')
  })
})
