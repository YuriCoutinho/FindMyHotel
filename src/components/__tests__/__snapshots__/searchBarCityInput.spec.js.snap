// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchBarCityInput.vue > Should match the snapshot 1`] = `
"<div class="p-autocomplete p-component p-inputwrapper" style="position: relative;" data-pc-name="autocomplete" data-pc-section="root" pc3=""><input type="text" class="p-inputtext p-component p-autocomplete-input w-full md:w-40 text-gray-700 font-medium border rounded-lg px-3 py-2 focus:outline-none focus:ring-1 focus:ring-emerald-500" placeholder="Cidade de destino" tabindex="0" autocomplete="off" role="combobox" aria-haspopup="listbox" aria-autocomplete="list" aria-expanded="false" aria-controls="undefined_panel" data-pc-name="pcinput" data-pc-extend="inputtext" data-pc-section="root" value="">
  <!---->
  <!---->
  <!----><span role="status" aria-live="polite" class="p-hidden-accessible" data-pc-section="hiddensearchresult" data-p-hidden-accessible="true">No results found</span>
  <!---->
</div>"
`;
